ui/StatusQ/src/keychain_osx.mm

```diff
@@ -0,0 +1,239 @@
+#include "StatusQ/keychain.h"
+
+#include <QDebug>
+#include <QEventLoop>
+#include <QFuture>
+#include <QtConcurrent/QtConcurrent>
+
+#include <Foundation/Foundation.h>
+#include <LocalAuthentication/LocalAuthentication.h>
+#include <Security/Security.h>
+
+const static auto authPolicy = LAPolicyDeviceOwnerAuthentication;
+
+static Keychain::Status convertStatus(OSStatus status)
+{
+    switch (status) {
+    case errSecSuccess:
+        return Keychain::StatusSuccess;
+    case errSecItemNotFound:
+        return Keychain::StatusNotFound;
+    case errSecCSCancelled:
+        return Keychain::StatusCancelled;
+    default:
+        return Keychain::StatusGenericError;
+    }
+}
+
+Keychain::Status convertError(NSError *error)
+{
+    switch (error.code) {
+    case errSecSuccess:
+        return Keychain::StatusSuccess;
+    case LAErrorSystemCancel:
+    case LAErrorUserCancel:
+    case LAErrorAppCancel:
+        return Keychain::StatusCancelled;
+    default:
+        return Keychain::StatusGenericError;
+    }
+}
+
+Keychain::~Keychain()
+{
+    cancelActiveRequest();
+    m_future.waitForFinished();
+}
+
+Keychain::Status authenticate(QString &reason, LAContext **context)
+{
+    if (context == nullptr)
+        return Keychain::StatusGenericError;
+
+    if (*context != nullptr) {
+        qWarning() << "another local authentication request in progress";
+        return Keychain::StatusGenericError;
+    }
+
+    *context = [[LAContext alloc] init];
+    NSError *authError = nil;
+
+    // Check if Biometrics Authentication is available
+    if (![*context canEvaluatePolicy:authPolicy error:&authError]) {
+        qWarning() << "biometric authentication not available:"
+                   << QString::fromNSString(authError.localizedDescription);
+        return convertError(authError);
+    }
+
+    QEventLoop loop;
+    auto loopPtr = &loop;
+    __block NSError *callbackError = nil;
+    __block BOOL success = NO;
+
+    // Prompt for biometrics authentication
+    [*context evaluatePolicy:authPolicy
+             localizedReason:reason.toNSString()
+                       reply:^(BOOL authSuccess, NSError *error) {
+                           success = authSuccess;
+                           callbackError = error ? [error copy] : nil;
+                           loopPtr->quit();
+                       }];
+
+    // Wait for biometrics authentication finished
+    loop.exec();
+
+    if (!success && callbackError) {
+        qWarning() << "authentication failed:"
+                   << QString::fromNSString(callbackError.localizedDescription);
+        return convertError(callbackError);
+    }
+
+    return Keychain::StatusSuccess;
+}
+
+void Keychain::requestSaveCredential(const QString &account, const QString &password)
+{
+    if (m_future.isRunning()) {
+        return;
```

Comment:
I think we should print a warning at least at this point. Calling it when busy means some logical problems on the caller side and shouldn't be just muted.